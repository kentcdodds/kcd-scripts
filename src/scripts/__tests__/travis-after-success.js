import cases from 'jest-in-case'

// this removes the quotes around strings...
expect.addSnapshotSerializer({
  print: val => val,
  test: val => typeof val === 'string',
})

cases(
  'travis-after-success',
  ({
    version = '0.0.0-semantically-released',
    hasCoverageDir = true,
    env = {
      TRAVIS: 'true',
      TRAVIS_BRANCH: 'master',
      TRAVIS_PULL_REQUEST: 'false',
    },
    runsNothing = false,
  }) => {
    // beforeEach
    const {sync: crossSpawnSyncMock} = require('cross-spawn')
    const utils = require('../../utils')
    utils.resolveBin = (modName, {executable = modName} = {}) => executable
    const originalEnvs = Object.keys(env).map(envKey => {
      const orig = process.env[envKey]
      process.env[envKey] = env[envKey]
      return orig
    })
    const originalLog = console.log
    const originalExit = process.exit
    process.exit = jest.fn()
    console.log = jest.fn()

    // tests
    crossSpawnSyncMock.mockClear()
    if (version) {
      utils.pkg.version = version
    }
    utils.hasFile = () => hasCoverageDir
    require('../travis-after-success')
    if (runsNothing) {
      expect(console.log.mock.calls).toMatchSnapshot()
    } else {
      expect(crossSpawnSyncMock).toHaveBeenCalledTimes(1)
      const [firstCall] = crossSpawnSyncMock.mock.calls
      const [script, calledArgs] = firstCall
      expect([script, ...calledArgs].join(' ')).toMatchSnapshot()
    }

    // afterEach
    process.exit = originalExit
    console.log = originalLog
    Object.keys(originalEnvs).forEach(envKey => {
      process.env[envKey] = env[envKey]
    })
    jest.resetModules()
  },
  {
    'calls concurrently with both scripts when on travis': {},
    'does not do the autorelease script when the version is different': {
      version: '1.2.3',
    },
    'does not do the codecov script when there is no coverage directory': {
      hasCoverageDir: false,
    },
    'does not do autorelease script when running on travis but in a pull request': {
      env: {
        TRAVIS: 'true',
        TRAVIS_BRANCH: 'master',
        TRAVIS_PULL_REQUEST: 'true',
      },
    },
    'does not run either script when no coverage dir and not the right version': {
      runsNothing: true,
      hasCoverageDir: false,
      version: '1.2.3',
    },
  },
)
