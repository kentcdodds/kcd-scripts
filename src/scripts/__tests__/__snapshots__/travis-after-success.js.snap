// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`travis-after-success calls concurrently with both scripts when on travis 1`] = `
Array [
  Array [
    installing and running travis-deploy-once,
  ],
]
`;

exports[`travis-after-success calls concurrently with both scripts when on travis 2`] = `
Array [
  npx travis-deploy-once@5,
  concurrently --prefix [{name}] --names codecov,release --prefix-colors bgBlue.bold.reset,bgGreen.bold.reset "echo installing codecov && npx -p codecov@3 -c 'echo running codecov && codecov'" "echo installing semantic-release && npx -p semantic-release@16 -c 'echo running semantic-release && semantic-release'",
]
`;

exports[`travis-after-success does not do autorelease script when running on travis but in a pull request 1`] = `Array []`;

exports[`travis-after-success does not do autorelease script when running on travis but in a pull request 2`] = `
Array [
  concurrently --prefix [{name}] --names codecov --prefix-colors bgBlue.bold.reset "echo installing codecov && npx -p codecov@3 -c 'echo running codecov && codecov'",
]
`;

exports[`travis-after-success does not do the autorelease script when the version is different 1`] = `Array []`;

exports[`travis-after-success does not do the autorelease script when the version is different 2`] = `
Array [
  concurrently --prefix [{name}] --names codecov --prefix-colors bgBlue.bold.reset "echo installing codecov && npx -p codecov@3 -c 'echo running codecov && codecov'",
]
`;

exports[`travis-after-success does not do the codecov script when opted out 1`] = `
Array [
  Array [
    installing and running travis-deploy-once,
  ],
]
`;

exports[`travis-after-success does not do the codecov script when opted out 2`] = `
Array [
  npx travis-deploy-once@5,
  concurrently --prefix [{name}] --names release --prefix-colors bgBlue.bold.reset "echo installing semantic-release && npx -p semantic-release@16 -c 'echo running semantic-release && semantic-release'",
]
`;

exports[`travis-after-success does not do the codecov script when there is no coverage directory 1`] = `
Array [
  Array [
    installing and running travis-deploy-once,
  ],
]
`;

exports[`travis-after-success does not do the codecov script when there is no coverage directory 2`] = `
Array [
  npx travis-deploy-once@5,
  concurrently --prefix [{name}] --names release --prefix-colors bgBlue.bold.reset "echo installing semantic-release && npx -p semantic-release@16 -c 'echo running semantic-release && semantic-release'",
]
`;

exports[`travis-after-success does not run either script when no coverage dir and not the right version 1`] = `
Array [
  Array [
    No need to autorelease or report coverage. Skipping travis-after-success script...,
  ],
]
`;

exports[`travis-after-success does not run either script when no coverage dir and not the right version 2`] = `Array []`;
